packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js

```diff
@@ -765,6 +765,45 @@ describe('ReactHooks', () => {
     }).toThrow('is not a function');
   });
 
+  it('assumes useEffect clean-up function is either a function or undefined', () => {
+    const {useEffect} = React;
+
+    function App(props) {
+      useEffect(() => {
+        return props.return;
+      });
+      return null;
+    }
+
+    const root1 = ReactTestRenderer.create(null);
+    expect(() => act(() => root1.update(<App return={17} />))).toErrorDev([
+      'Warning: An effect function must not return anything besides a ' +
+        'function, which is used for clean-up. You returned: 17',
+    ]);
+
+    const root2 = ReactTestRenderer.create(null);
+    expect(() => act(() => root2.update(<App return={null} />))).toErrorDev([
+      'Warning: An effect function must not return anything besides a ' +
+        'function, which is used for clean-up. You returned null. If your ' +
+        'effect does not require clean up, return undefined (or nothing).',
+    ]);
+
+    const root3 = ReactTestRenderer.create(null);
+    expect(() =>
+      act(() => root3.update(<App return={Promise.resolve()} />)),
+    ).toErrorDev([
+      'Warning: An effect function must not return anything besides a ' +
+        'function, which is used for clean-up.\n\n' +
+        'It looks like you wrote useEffect(async () => ...) or returned a Promise.',
+    ]);
+
+    // TODO: we're not checking that the passive destroy function
+    // is a function like we do for layout effects in the test above.
+    expect(() => act(() => root3.update(null))).toErrorDev(
+      'func.apply is not a function',
```

Comment:
This probably should throw an error, like how `useLayoutEffect` does, right?

Reply:
Yeah, that's what the TODO says. Was thinking of passing this to someone as a good first issue?